### PR TITLE
WIP: Cryptofuzz tests integration

### LINKED ISSUE #9744: Cryptofuzz in OpenSSL CI
As the master branch keeps receiving many updates, new bugs keep getting found in Cryptofuzz/OSS-Fuzz.

I think the quality of your software could benefit from a CI integration of Cryptofuzz. Cryptofuzz + a static corpus (eg. never or infrequently updated) and without differential capabilities (eg. don't test against other libraries) can detect many bugs as soon as they are attempted to be merged.

Just today 2 new bugs were found:

- A UAF https://oss-fuzz.com/testcase-detail/5658978675064832
- Invalid read https://oss-fuzz.com/testcase-detail/5696325059870720

(might be related to one another)

These are probably due to some recent change, and if Cryptofuzz was part of your CI tests, you might have been alerted sooner.

How do you feel about this?

### PR DESCRIPTION
**Pull request by guidovranken**
Fixes #9744

##### Checklist
- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: test/recipes/95-test_external_cryptofuzz.t [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+
+setup("test_external_cryptofuzz");
+
+plan skip_all => "No external tests in this configuration"
+    if disabled("external-tests");
+plan skip_all => "Cryptofuzz not available"
+    if ! -f srctop_file("cryptofuzz", "README.md");
+plan skip_all => "Cryptofuzz only available in a static build"
+    if disabled("static");
```
Comment (levitte) [inline test/recipes/95-test_external_cryptofuzz.t]: There should be a similar skip_all if `$^O` is 'VMS' or 'MSWin32'

**Comment by levitte** [comment]
This is looking pretty good.  You're missing something, though, according to Travis:

``` console
95-test_external_cryptofuzz.t .. 1/1 
1..1
make[2]: Entering directory '/home/travis/build/openssl/openssl/cryptofuzz/modules/openssl'
In file included from ../../fuzzing-headers/include/fuzzing/types.hpp:7:0,
                 from ../../fuzzing-headers/include/fuzzing/datasource/datasource.hpp:5,
                 from ../../include/cryptofuzz/generic.h:5,
                 from ../../include/cryptofuzz/components.h:3,
                 from module.h:3,
                 from module.cpp:1:
../../fuzzing-headers/include/fuzzing/memory.hpp:5:20: fatal error: optional: No such file or directory
compilation terminated.
Makefile:10: recipe for target 'module.o' failed
make[2]: *** [module.o] Error 1
make[2]: Leaving directory '/home/travis/build/openssl/openssl/cryptofuzz/modules/openssl'
make[2]: Entering directory '/home/travis/build/openssl/openssl/cryptofuzz'
In file included from fuzzing-headers/include/fuzzing/types.hpp:7:0,
                 from fuzzing-headers/include/fuzzing/datasource/datasource.hpp:5,
                 from driver.h:5,
                 from driver.cpp:1:
fuzzing-headers/include/fuzzing/memory.hpp:5:20: fatal error: optional: No such file or directory
compilation terminated.
Makefile:6: recipe for target 'driver.o' failed
make[2]: *** [driver.o] Error 1
make[2]: Leaving directory '/home/travis/build/openssl/openssl/cryptofuzz'
/home/travis/build/openssl/openssl/util/shlib_wrap.sh sh /home/travis/build/openssl/openssl
test/recipes/95-test_external_cryptofuzz_data/cryptofuzz.sh => 2
not ok 1 - running Cryptofuzz tests
```

**Comment by guidovranken** [comment]
> This is looking pretty good. You're missing something, though, according to Travis:

Yeah, the problem is that Cryptofuzz uses the [C++ 17 optional header](https://en.cppreference.com/w/cpp/utility/optional), and older versions of GCC support it by providing ```<experimental/optional>``` instead of ```<optional>```.

I think I can resolve this on the fly with ```sed```.

However, there is another issue ahead: GCC does not have libFuzzer built-in, while (recent) clang does (with ```-fsanitize=fuzzer```). Of course we are not going to fuzz in Travis, but I need libFuzzer as a driver to load and run inputs in the corpus. Would it be possible to compile the external tests, or at least Cryptofuzz, with clang? This might also spontaneously resolve the issue with ```<optional>```.

Furthermore, for this effort to be as effective as possible, I'd prefer to compile OpenSSL with ```-fsanitize=address,undefined```. Are you OK with this, and should I either:

- Alter the existing rules in ```.travis.yml``` to use clang + sanitizers
- Create a separate target for Cryptofuzz

**Comment by levitte** [comment]
I have zero problems switching that job to use clang.

**Comment by kroeckx** [comment]
We have a fuzz/test-corpus.c file that we use, is that useful for you?

**Comment by guidovranken** [comment]
@kroeckx I had to use clang to work around the <optional> issue, and it turns out the version of clang that Travis uses is recent enough to support ```-fsanitize=fuzzer```, so that problem is solved. Thanks though.

The build seems to be succeeding (Travis signals failure because the Pyca test segfaults, I think). The next steps are adding a corpus (I'll add this as a submodule to Cryptofuzz, so it gets downloaded automatically), and running Cryptofuzz against this corpus.

What do you think about the sanitizers? I can add a new target specifically for Cryptofuzz, or add sanitizers to the existing build rule for external tests. The latter alternative will unnecessarily slow down the building/running of the existing tests though. Let me know, thanks.

**Comment by richsalz** [comment]
I don't think we want to slow down the existing tests.

**Comment by kroeckx** [comment]
We already have a target with ubsan, asan and msan, I don't see a
problem with an other target enabling them. I'm also fine with the
other external tests running with them enabled.

Maybe it's useful that not all external tests get run in the same
travis job.

**Comment by guidovranken** [comment]
OK, I'll modify the script now to build with the sanitizers. We can always revert and create a separate rule in the final stage.

> Maybe it's useful that not all external tests get run in the same travis job.

Agreed, especially if there are one or more broken tests (as seems to be the case with Pyca). The Travis job will fail regardless of whether a PR contains bugs.

**Comment by t8m** [comment]
Could you please rebase the PR to resolve conflicts?

Marking as inactive to be closed at the end of 3.4 dev barring further input.

**Comment by richsalz** [comment]
@guidovranken your toolkit has found bugs before, right?  interested in bring it to quictls?

**Pull request closed without merging**